include/cetl/any.hpp

```diff
@@ -0,0 +1,766 @@
+/// @file
+/// Defines the C++17 `std::any` type and several related entities.
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef CETL_ANY_HPP_INCLUDED
+#define CETL_ANY_HPP_INCLUDED
+
+#include "rtti.hpp"
+#include "pf17/cetlpf.hpp"
+#include "pf17/utility.hpp"
+#include "pf17/attribute.hpp"
+
+#include <algorithm>
+#include <initializer_list>
+
+#ifndef CETL_H_ERASE
+#    include "cetl/cetl.hpp"
+#endif
+
+namespace cetl
+{
+
+// Forward declarations
+template <std::size_t Footprint, bool Copyable, bool Movable, std::size_t Alignment>
+class any;
+
+namespace detail
+{
+
+template <std::size_t Footprint, std::size_t Alignment>
+struct base_storage  // NOLINT(*-pro-type-member-init)
+{
+    base_storage() = default;
+
+    CETL_NODISCARD void* get_raw_storage() noexcept
+    {
+        return static_cast<void*>(buffer_);
+    }
+
+    CETL_NODISCARD const void* get_raw_storage() const noexcept
+    {
+        return static_cast<const void*>(buffer_);
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return nullptr != value_destroyer_;
+    }
+
+    template <typename Tp>
+    void make_handlers() noexcept
+    {
+        static_assert(sizeof(Tp) <= Footprint, "Enlarge the footprint");
+
+        CETL_DEBUG_ASSERT(nullptr == value_destroyer_, "Expected to be empty before making handlers.");
+        CETL_DEBUG_ASSERT(nullptr == value_converter_, "");
+        CETL_DEBUG_ASSERT(nullptr == value_const_converter_, "");
+
+        value_destroyer_ = [](void* const storage) {
+            const auto ptr = static_cast<Tp*>(storage);
+            ptr->~Tp();
+        };
+
+        make_converters<Tp>();
+    }
+
+    template <typename Tp, std::enable_if_t<is_rtti_convertible<Tp>, int> = 0>
+    void make_converters() noexcept
+    {
+        value_const_converter_ = [](const void* const storage, const type_id& id) {
+            const auto ptr = static_cast<const Tp*>(storage);
+            return ptr->_cast_(id);
+        };
+        value_converter_ = [](void* const storage, const type_id& id) {
+            auto ptr = static_cast<Tp*>(storage);
+            return ptr->_cast_(id);
+        };
+    }
+
+    template <typename Tp, std::enable_if_t<!is_rtti_convertible<Tp>, int> = 0>
+    void make_converters() noexcept
+    {
+        value_const_converter_ = [](const void* const storage, const type_id& id) {
+            return (id == type_id_value<Tp>) ? storage : nullptr;
```

Comment:
We need to change `type_id_value` to be functional to allow for `typeid`'s required runtime. That is, the token must be replaceable with typeid when CETL RTTI is erased. Here's an example polyfill:

```
template <typename T> inline constexpr const type_id &type_id_value()
{
    return typeid(T);
}
```

This would require the token to look like `type_id_value<Tp>()` so we can invoke `typeid`.

Reply:
@thirtytwobits Scott, is it possible that we handle this request as a separate issue? I'm asking b/c I currently don't understand requirements, as well as the **exact** way to reproduce it.

Reply:
ok, I did the replacement. unfortunately, I still don't understand what exactly was the problem (and how to reproduce it), but I did as you proposed